src/java/org/apache/cassandra/db/marshal/AbstractType.java

```diff
@@ -415,29 +453,77 @@ public boolean isVector()
         return false;
     }
 
-    public boolean isMultiCell()
+    public boolean isFreezable()
     {
         return false;
     }
 
-    public boolean isFreezable()
+    /**
+     * If the type is a multi-cell one ({@link #isMultiCell} is true), returns a frozen copy of this type (one
+     * for which {@link #isMultiCell} returns false).
+     * <p>
+     * Note that as mentioned on {@link #isMultiCell}, a frozen type necessarily has all its subtypes frozen, so
+     * this method also ensures that no subtypes (recursively) are marked as multi-cell.
+     *
+     * @return a frozen version of this type. If this type is not multi-cell (whether because it is not a "complex"
+     * type, or because it is already a frozen one), this should return {@code this}.
+     */
+    public AbstractType<?> freeze()
     {
-        return false;
+        if (!isMultiCell)
+            return this;
+
+        return with(freeze(subTypes), false);
     }
 
-    public AbstractType<?> freeze()
+    /**
+     * Utility method that freezes a list of types.
+     *
+     * @param types the list of types to freeze.
+     * @return a new (unmodifiable) list containing the result of applying {@link #freeze()} on every type of
+     * {@code types}.
+     */
+    public static List<AbstractType<?>> freeze(List<AbstractType<?>> types)
     {
-        return this;
+        return types.stream().map(AbstractType::freeze).collect(Collectors.toUnmodifiableList());
     }
 
-    public AbstractType<?> unfreeze()
+    /**
+     * Creates an instance of this type (the concrete type extending this class) with the provided updated multi-cell
+     * flag and subtypes.
+     * <p>
+     * Any other information (other than multi-cellness and subtypes) the type may have is expected to be left unchanged
+     * in the created type.
+     *
+     * @param isMultiCell whether the returned type must be a multi-cell one or not.
+     * @param subTypes the subtypes to use for the returned type as a list. The list will have subtypes in the exact
+     * same order as returned by {@link #subTypes}, and exactly as many as the concrete class expects.
+     * @return the created type, which can be {@code this} if the provided subTypes and multi-cell flag are the same
+     * as that of this type.
+     */
+    public AbstractType<?> with(List<AbstractType<?>> subTypes, boolean isMultiCell)
     {
+        // Default implementation for types that can neither be multi-cell, nor have subtypes (and thus where this
+        // is basically a no-op). Any other type must override this.
+
+        assert this.subTypes.isEmpty() && subTypes.isEmpty() :
+        String.format("Invalid call to copyWith on %s with subTypes %s (provided subTypes: %s)",
+                      this, this.subTypes, subTypes);
+
+        assert !this.isMultiCell && !isMultiCell:
+        String.format("Invalid call to copyWith on %s with isMultiCell %b (provided isMultiCell: %b)",
```

Comment:
It seems the original path at some point named the method `copyWith` instead of `with`:
```suggestion
        String.format("Invalid call to with on %s with subTypes %s (provided subTypes: %s)",
                      this, this.subTypes, subTypes);

        assert !this.isMultiCell && !isMultiCell:
        String.format("Invalid call to with on %s with isMultiCell %b (provided isMultiCell: %b)",
```